File: gravitee-apim-definition/gravitee-apim-definition-model/src/main/java/io/gravitee/definition/model/Proxy.java
/*
 * Copyright © 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.gravitee.definition.model;

import com.fasterxml.jackson.annotation.*;
import java.io.Serializable;
import java.util.*;
import java.util.stream.Collectors;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.NoArgsConstructor;

/**
 * @author David BRASSELY (david.brassely at graviteesource.com)
 * @author GraviteeSource Team
 */
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Proxy implements Serializable {

    @JsonProperty("virtual_hosts")
    private List<VirtualHost> virtualHosts;

    @JsonProperty("groups")
    private Set<EndpointGroup> groups;

    @JsonProperty("failover")
    private Failover failover;

    @JsonProperty("cors")
    private Cors cors;

    @JsonProperty("logging")
    private Logging logging;

    @JsonProperty("strip_context_path")
    @Builder.Default
    private boolean stripContextPath = false;

    @JsonProperty("preserve_host")
    @Builder.Default
    private boolean preserveHost = false;

    @JsonProperty("servers")
    private List<String> servers;

    public boolean isStripContextPath() {
        return stripContextPath;
    }

    public void setStripContextPath(boolean stripContextPath) {
        this.stripContextPath = stripContextPath;
    }

    public List<VirtualHost> getVirtualHosts() {
        return virtualHosts;
    }

    public void setVirtualHosts(List<VirtualHost> virtualHosts) {
        this.virtualHosts = virtualHosts;
    }

    public boolean failoverEnabled() {
        return failover != null;
    }

    public Failover getFailover() {
        return failover;
    }

    public void setFailover(Failover failover) {
        this.failover = failover;
    }

    public Cors getCors() {
        return cors;
    }

    public void setCors(Cors cors) {
        this.cors = cors;
    }

    public Set<EndpointGroup> getGroups() {
        return groups;
    }

    public void setGroups(Set<EndpointGroup> groups) {
        this.groups = groups;
    }

    public Logging getLogging() {
        return logging;
    }

    public void setLogging(Logging logging) {
        this.logging = logging;
    }

    public boolean isPreserveHost() {
        return preserveHost;
    }

    public void setPreserveHost(boolean preserveHost) {
        this.preserveHost = preserveHost;
    }

    public List<String> getServers() {
        return servers;
    }

    public void setServers(List<String> servers) {
        this.servers = servers;
    }

    @JsonIgnore
    public String getContextPath() {
        if (this.virtualHosts == null || this.virtualHosts.isEmpty()) {
            return null;
        }
        return this.virtualHosts.get(0).getPath();
    }

    @JsonIgnore
    public List<Plugin> getPlugins() {
        return Optional.ofNullable(this.groups)
            .map(g -> g.stream().map(EndpointGroup::getPlugins).flatMap(List::stream).collect(Collectors.toList()))
            .orElse(List.of());
    }
}

File: gravitee-apim-perf/src/lib/models/v3/Proxy.ts
/*
 * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { EndpointGroup } from '@lib/models/v3/EndpointGroup';

export interface Proxy {
  /**
   *
   * @type {Cors}
   * @memberof Proxy
   */
  cors?: Cors;
  /**
   *
   * @type {Failover}
   * @memberof Proxy
   */
  failover?: Failover;
  /**
   *
   * @type {Array<EndpointGroup>}
   * @memberof Proxy
   */
  groups?: Array<EndpointGroup>;
  /**
   *
   * @type {Logging}
   * @memberof Proxy
   */
  logging?: Logging;
  /**
   *
   * @type {boolean}
   * @memberof Proxy
   */
  preserve_host?: boolean;
  /**
   *
   * @type {boolean}
   * @memberof Proxy
   */
  strip_context_path?: boolean;
  /**
   *
   * @type {Array<VirtualHost>}
   * @memberof Proxy
   */
  virtual_hosts?: Array<VirtualHost>;
}

export interface Cors {
  /**
   *
   * @type {boolean}
   * @memberof Cors
   */
  allowCredentials?: boolean;
  /**
   *
   * @type {Array<string>}
   * @memberof Cors
   */
  allowHeaders?: Array<string>;
  /**
   *
   * @type {Array<string>}
   * @memberof Cors
   */
  allowMethods?: Array<string>;
  /**
   *
   * @type {Array<string>}
   * @memberof Cors
   */
  allowOrigin?: Array<string>;
  /**
   *
   * @type {boolean}
   * @memberof Cors
   */
  enabled?: boolean;
  /**
   *
   * @type {Array<string>}
   * @memberof Cors
   */
  exposeHeaders?: Array<string>;
  /**
   *
   * @type {number}
   * @memberof Cors
   */
  maxAge?: number;
  /**
   *
   * @type {boolean}
   * @memberof Cors
   */
  runPolicies?: boolean;
}

export interface Failover {
  /**
   *
   * @type {Array<string>}
   * @memberof Failover
   */
  cases?: Array<FailoverCasesEnum>;
  /**
   *
   * @type {number}
   * @memberof Failover
   */
  maxAttempts?: number;
  /**
   *
   * @type {number}
   * @memberof Failover
   */
  retryTimeout?: number;
}

export interface Logging {
  /**
   *
   * @type {string}
   * @memberof Logging
   */
  condition?: string;
  /**
   *
   * @type {string}
   * @memberof Logging
   */
  content?: LoggingContentEnum;
  /**
   *
   * @type {string}
   * @memberof Logging
   */
  mode?: LoggingModeEnum;
  /**
   *
   * @type {string}
   * @memberof Logging
   */
  scope?: LoggingScopeEnum;
}

export interface VirtualHost {
  /**
   *
   * @type {string}
   * @memberof VirtualHost
   */
  host?: string;
  /**
   *
   * @type {boolean}
   * @memberof VirtualHost
   */
  override_entrypoint?: boolean;
  /**
   *
   * @type {string}
   * @memberof VirtualHost
   */
  path?: string;
}

export const enum FailoverCasesEnum {
  TIMEOUT = 'TIMEOUT',
}

export const enum LoggingContentEnum {
  NONE = 'NONE',
  HEADERS = 'HEADERS',
  PAYLOADS = 'PAYLOADS',
  HEADERS_PAYLOADS = 'HEADERS_PAYLOADS',
}

/**
 * @export
 */
export const enum LoggingModeEnum {
  NONE = 'NONE',
  CLIENT = 'CLIENT',
  PROXY = 'PROXY',
  CLIENT_PROXY = 'CLIENT_PROXY',
}

/**
 * @export
 */
export const enum LoggingScopeEnum {
  NONE = 'NONE',
  REQUEST = 'REQUEST',
  RESPONSE = 'RESPONSE',
  REQUEST_RESPONSE = 'REQUEST_RESPONSE',
}
